Incompatibility with Maven 4

Hi,
I have a project (Apache Commons Fileuoload), which uses the spdx-maven-plugin, version 1.0.2. (Latest, as of this writing.)
On Windows 111, this orks perfectly fine with Maven 3, but not with Maven 4. Steps to reproduce:

```
$ git clone https://gitbox.apache.org/repos/asf/commons-fileupload.git
$ cd commons-fileupload
# Ignore warnings for detached head, etc. in the following step:
$ git checkout f3f27401baea7e8a0be8155fa26afe026d0205af
# With Maven 3, the build works just fine.
$ /c/opt/mvn/3/bin/mvn
# With Maven 4, an error occurs:
$ /c/opt/mvn/4/bin/mvn -e
```

The error message:

```
[ERROR] Failed to execute goal org.spdx:spdx-maven-plugin:1.0.2:createSPDX (build-sbom-spdx) on project commons-fileupload2-distribution: Execution build-sbom-spdx of goal org.spdx:spdx-maven-plugin:1.0.2:createSPDX failed: Range [87, 65) out of bounds for length 65 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.spdx:spdx-maven-plugin:1.0.2:createSPDX (build-sbom-spdx) on project commons-fileupload2-distribution: Execution build-sbom-spdx of goal org.spdx:spdx-maven-plugin:1.0.2:createSPDX failed: Range [87, 65) out of bounds for length 65
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:346)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:310)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:110)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:61)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleStarter.execute(DefaultLifecycleStarter.java:123)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:310)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:449)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke(Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution build-sbom-spdx of goal org.spdx:spdx-maven-plugin:1.0.2:createSPDX failed: Range [87, 65) out of bounds for length 65
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:152)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:339)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:310)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:110)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:61)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleStarter.execute(DefaultLifecycleStarter.java:123)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:310)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:449)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke(Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
Caused by: java.lang.StringIndexOutOfBoundsException: Range [87, 65) out of bounds for length 65
    at jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
    at jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
    at jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
    at jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
    at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
    at jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:112)
    at jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:349)
    at java.lang.String.checkBoundsBeginEnd(String.java:4865)
    at java.lang.String.substring(String.java:2834)
    at java.lang.String.substring(String.java:2807)
    at org.spdx.maven.utils.SpdxV2FileCollector.collectFiles(SpdxV2FileCollector.java:100)
    at org.spdx.maven.utils.SpdxV2DocumentBuilder.collectSpdxFileInformation(SpdxV2DocumentBuilder.java:375)
    at org.spdx.maven.CreateSpdxMojo.execute(CreateSpdxMojo.java:626)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:146)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:339)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:310)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:110)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:61)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleStarter.execute(DefaultLifecycleStarter.java:123)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:310)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:449)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke(Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
[ERROR]
[ERROR] Re-run Maven using the '-X' switch to enable verbose output
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn [args] -r
```
 
Maven 4 information:
```
jwi@Lap-Jochen-Win MINGW64 /c/Work/git/gitbox.apache.org/commons-fileupload-orig ((f3f27401...))
$ /c/opt/mvn/4/bin/mvn --version
Apache Maven 4.0.0-rc-4 (bed0f8174bf728978f86fac533aa38a9511f3872)
Maven home: C:\opt\mvn\4.0.0-rc-4
Java version: 21.0.7, vendor: Eclipse Adoptium, runtime: C:\opt\jdk\21.0.7.6
Default locale: en_US, platform encoding: UTF-8
OS name: "windows 11", version: "10.0", arch: "amd64", family: "winnt"
```

--- a/src/main/java/org/spdx/maven/CreateSpdxMojo.java
+++ b/src/main/java/org/spdx/maven/CreateSpdxMojo.java
@@ -120,7 +120,7 @@ public class CreateSpdxMojo extends AbstractMojo
         SpdxModelFactory.init();
     }
 
-    @Component
+    @Parameter( defaultValue = "${project}", readonly = true )
     private MavenProject mavenProject;
 
     @Component
@@ -129,7 +129,7 @@ public class CreateSpdxMojo extends AbstractMojo
     @Component
     protected ProjectBuilder mavenProjectBuilder;
 
-    @Component
+    @Parameter( defaultValue = "${session}", readonly = true )
     protected MavenSession session;
 
     @Component(hint = "default")

--- a/src/main/java/org/spdx/maven/utils/AbstractFileCollector.java
+++ b/src/main/java/org/spdx/maven/utils/AbstractFileCollector.java
@@ -8,6 +8,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.util.Collection;
@@ -233,6 +234,15 @@ public abstract class AbstractFileCollector
         }
 
         return checksums;
+    }
+
+    /**
+     * @param file file which is the target for the path string
+     * @param baseDir base directory of the SPDX archive or project containing the file
+     * @return a file path for the file relative to the baseDir
+     */
+    protected static String toRelativeFilePath( File file, String baseDir ) {
+        return Path.of( baseDir ).relativize( file.toPath() ).toString().replace( '\\', '/' );
     }
 
 }

--- a/src/main/java/org/spdx/maven/utils/SpdxV2FileCollector.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV2FileCollector.java
@@ -97,23 +97,15 @@ public class SpdxV2FileCollector extends AbstractFileCollector
             {
                 String filePath = fileSet.getDirectory() + File.separator + includedFile;
                 File file = new File( filePath );
-                String relativeFilePath = file.getAbsolutePath().substring( baseDir.length() + 1 ).replace( '\\', '/' );
+                String relativeFilePath = toRelativeFilePath( file, baseDir );
                 SpdxDefaultFileInformation fileInfo = findDefaultFileInformation( relativeFilePath,
                         pathSpecificInformation );
                 if ( fileInfo == null )
                 {
                     fileInfo = defaultFileInformation;
                 }
-
-                String outputFileName;
-                if ( fileSet.getOutputDirectory() != null )
-                {
-                    outputFileName = fileSet.getOutputDirectory() + File.separator + includedFile;
-                }
-                else
-                {
-                    outputFileName = file.getAbsolutePath().substring( baseDir.length() + 1 );
-                }
+                String outputFileName = Objects.nonNull( fileSet.getOutputDirectory() ) ?
+                        fileSet.getOutputDirectory() + File.separator + includedFile : relativeFilePath;
                 collectFile( file, outputFileName, fileInfo, relationshipType, projectPackage, spdxDoc, algorithms );
             }
         }

--- a/src/main/java/org/spdx/maven/utils/SpdxV3FileCollector.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV3FileCollector.java
@@ -130,23 +130,15 @@ public class SpdxV3FileCollector extends AbstractFileCollector
             {
                 String filePath = fileSet.getDirectory() + File.separator + includedFile;
                 File file = new File( filePath );
-                String relativeFilePath = file.getAbsolutePath().substring( baseDir.length() + 1 ).replace( '\\', '/' );
+                String relativeFilePath = toRelativeFilePath( file, baseDir );
                 SpdxDefaultFileInformation fileInfo = findDefaultFileInformation( relativeFilePath,
                         pathSpecificInformation );
                 if ( fileInfo == null )
                 {
                     fileInfo = defaultFileInformation;
                 }
-
-                String outputFileName;
-                if ( fileSet.getOutputDirectory() != null )
-                {
-                    outputFileName = fileSet.getOutputDirectory() + File.separator + includedFile;
-                }
-                else
-                {
-                    outputFileName = file.getAbsolutePath().substring( baseDir.length() + 1 );
-                }
+                String outputFileName = Objects.nonNull( fileSet.getOutputDirectory() ) ?
+                        fileSet.getOutputDirectory() + File.separator + includedFile : relativeFilePath;
                 collectFile( file, outputFileName, fileInfo, relationshipType, projectPackage, spdxDoc, algorithms );
             }
         }

--- a/src/test/java/org/spdx/maven/utils/TestSpdxV2FileCollector.java
+++ b/src/test/java/org/spdx/maven/utils/TestSpdxV2FileCollector.java
@@ -674,9 +674,5 @@ public class TestSpdxV2FileCollector
         result = AbstractFileCollector.toRelativeFilePath( new File( "/ghi/file.java" ),
                 "/abc/def" );
         assertEquals( "../../ghi/file.java", result );
-        // test DOS filenames
-        result = AbstractFileCollector.toRelativeFilePath( new File( "c:\\abc\\def\\ghi\\file.java" ),
-                "c:\\abc\\def" );
-        assertEquals( "ghi/file.java", result );
     }
 }

--- a/src/test/java/org/spdx/maven/utils/TestSpdxV2FileCollector.java
+++ b/src/test/java/org/spdx/maven/utils/TestSpdxV2FileCollector.java
@@ -658,4 +658,25 @@ public class TestSpdxV2FileCollector
             }
         }
     }
+
+    @Test
+    public void testToRelativeFilePath()
+    {
+        // test simple case
+        String result = AbstractFileCollector.toRelativeFilePath( new File( "/abc/def/ghi/file.java" ),
+                "/abc/def" );
+        assertEquals( "ghi/file.java", result );
+        // test level up
+        result = AbstractFileCollector.toRelativeFilePath( new File( "/abc/file.java" ),
+                "/abc/def" );
+        assertEquals( "../file.java", result );
+        // test not relative
+        result = AbstractFileCollector.toRelativeFilePath( new File( "/ghi/file.java" ),
+                "/abc/def" );
+        assertEquals( "../../ghi/file.java", result );
+        // test DOS filenames
+        result = AbstractFileCollector.toRelativeFilePath( new File( "c:\\abc\\def\\ghi\\file.java" ),
+                "c:\\abc\\def" );
+        assertEquals( "ghi/file.java", result );
+    }
 }

--- a/src/test/java/org/spdx/maven/utils/TestSpdxV2LicenseManager.java
+++ b/src/test/java/org/spdx/maven/utils/TestSpdxV2LicenseManager.java
@@ -51,6 +51,7 @@ public class TestSpdxV2LicenseManager
     private static final String TEST_SPDX_DOCUMENT_URL = "http://www.spdx.org/documents/test";
     static final String APACHE_CROSS_REF_URL2 = "http://www.apache.org/licenses/LICENSE-2.0";
     static final String APACHE_CROSS_REF_URL3 = "http://opensource.org/licenses/Apache-2.0";
+    static final String APACHE_CROSS_REF_URL4 = "http://opensource.org/license/apache-2-0";
     static final String APACHE_LICENSE_ID = "Apache-2.0";
     static final String APACHE_LICENSE_NAME = "Apache License 2.0";
     static final String APSL_CROSS_REF_URL = "http://www.opensource.apple.com/source/IOSerialFamily/IOSerialFamily-7/APPLE_LICENSE";
@@ -277,7 +278,8 @@ public class TestSpdxV2LicenseManager
         License result = licenseManager.spdxLicenseToMavenLicense( licenseInfo );
         assertEquals( result.getName(), ( (SpdxListedLicense) licenseInfo ).getName() );
         String resultUrl = result.getUrl().replace( "https", "http" );
-        assertTrue( APACHE_CROSS_REF_URL2.equals( resultUrl ) || APACHE_CROSS_REF_URL3.equals( resultUrl ) );
+        assertTrue( APACHE_CROSS_REF_URL2.equals( resultUrl ) || APACHE_CROSS_REF_URL3.equals( resultUrl ) ||
+                APACHE_CROSS_REF_URL4.equals( resultUrl ) );
 
         // non standard license
         final String LICENSE_ID = "LicenseRef-nonStd1";

--- a/src/test/java/org/spdx/maven/utils/TestSpdxV3LicenseManager.java
+++ b/src/test/java/org/spdx/maven/utils/TestSpdxV3LicenseManager.java
@@ -54,6 +54,7 @@ public class TestSpdxV3LicenseManager
     private static final String TEST_SPDX_DOCUMENT_URL = "http://www.spdx.org/documents/test";
     static final String APACHE_CROSS_REF_URL2 = "http://www.apache.org/licenses/LICENSE-2.0";
     static final String APACHE_CROSS_REF_URL3 = "http://opensource.org/licenses/Apache-2.0";
+    static final String APACHE_CROSS_REF_URL4 = "http://opensource.org/license/apache-2-0";
     static final String APACHE_LICENSE_ID = "Apache-2.0";
     static final String APACHE_LICENSE_NAME = "Apache License 2.0";
     static final String APSL_CROSS_REF_URL = "http://www.opensource.apple.com/source/IOSerialFamily/IOSerialFamily-7/APPLE_LICENSE";
@@ -288,7 +289,8 @@ public class TestSpdxV3LicenseManager
         License result = licenseManager.spdxLicenseToMavenLicense( licenseInfo );
         assertEquals( result.getName(), ( (ListedLicense) licenseInfo ).getName().get() );
         String resultUrl = result.getUrl().replace( "https", "http" );
-        assertTrue( APACHE_CROSS_REF_URL2.equals( resultUrl ) || APACHE_CROSS_REF_URL3.equals( resultUrl ) );
+        assertTrue( APACHE_CROSS_REF_URL2.equals( resultUrl ) || APACHE_CROSS_REF_URL3.equals( resultUrl ) ||
+                APACHE_CROSS_REF_URL4.equals( resultUrl ) );
 
         // non standard license
         final String LICENSE_ID = "LicenseRef-nonStd1";

